docs/gcc_incompatibilities.md

```diff
@@ -0,0 +1,201 @@
+# GCC - Clang system headers incompatibilities
+
+## GCC fixincludes
+
+A little-known part of GCC's build process is a script called "fixincludes", or
+fixinc.sh. Purportedly, the purpose of this script is to fix "non-ANSI system
+header files" which GCC "cannot compile" (https://ewontfix.com/12/).
+Some examples of these fixes are:
+* Changing AIX's _LARGE_FILES redirection of open to open64, etc. to use GCC's
+  `__asm__` keyword rather than #define, as the latter breaks C++.
+* Exposing the long double math functions in math.h on Mac OS 10.3.9, which
+  inexplicably omitted declarations for them.
+* Adding workaround for Linux 2.2 and earlier kernel bug with direction flag to
+  FD_ZERO macros.
+
+The fixincludes process iterates over each header file it finds under
+/usr/include (or whatever the configured include directory is), applies a set
+of heuristics based on the filename, machine tuple, and a regular expression
+matched against the file contents, and if these rules match, it applies a
+sequence of sed commands to the file. As of this writing, there are 228 such
+"hacks" that may be applied. The output is then stored in GCC's private
+include-fixed directory (roughly /usr/lib/gcc/$MACH/$VER/include-fixed), which
+GCC searches before the system include directory, thus "replacing" the original
+header when the new GCC is used.
+
+Fixincludes is a bad idea and a bad hack, the [Gentoo Linux distribution simply
+removes the whole generated directory](https://sources.gentoo.org/cgi-bin/viewvc.cgi/gentoo-x86/eclass/toolchain.eclass?view=markup&sortby=log#l1524).
+Consequently, Clang (from version 3.8) skips the `include-fixed` include path
+even if it is built with GCC and configured to use GCC's libstdc++
+(https://bugs.launchpad.net/ubuntu/+source/llvm-toolchain-3.8/+bug/1573778).
+
+References:
+(https://www.gnu.org/software/autogen/fixinc.html)
+(https://android.googlesource.com/toolchain/gcc/+/master/gcc-4.8.3/fixincludes/README)
+(https://ewontfix.com/12/)
+
+## System headers for Intel vector instructions
+
+GCC and Clang support native vector operations differently.  The corresponding
+Intel intrinsics (for MMX, SSE, AVX, etc) have their own headers in a compiler
+specific include directory.  For instance, the `_mm_add_ss` for SSE can be
+found in xmmintrin.h.  GCC implements the intrinsic by delegating to it's own
+builtin intrinsic:
+```
```

Comment:
Use ` ```cpp` or ` ```c` to specify the language of the code. This will give us syntax highlight on the code.